File: src/states/layer2_state.py
"""Layer 2 – The Weighted Reward.

Player's A* is modified to route through a -10 step bonus reward that lies
slightly off the optimal direct path.  Dijkstra monster ignores it.
"""
import random
import pygame
from src.states.base_state import BaseState
from src.maze    import generate_square_maze, SquareGrid
from src.algorithms import AStarManhattan, Dijkstra
from src.entities   import Player, Monster
from src.ui.renderer import draw_square_maze, update_fog
from src.ui.sidebar  import draw_sidebar
from src.config import (
    GRID_COLS, GRID_ROWS, FOG_REVEAL_RADIUS,
    UI_BG, UI_SUCCESS, UI_TEXT, UI_SUBTEXT, UI_PANEL,
    ASTAR_MAN_COL, DIJKSTRA_COL, REWARD_COLOR,
    SCREEN_HEIGHT, MAZE_AREA_WIDTH, font,
)

_START  = (0, GRID_ROWS - 1)
_END    = (GRID_COLS - 1, 0)
_REWARD_BONUS = -10   # step penalty applied on collection

_INSTRUCT = [
    "Arrow keys to move",
    "Gold circle = -10 step reward",
    "Your path is routed via reward",
    "Dijkstra ignores it",
    "ESC → Menu",
]


class Layer2State(BaseState):

    def enter(self) -> None:
        seed          = random.randint(0, 9999)
        passages      = generate_square_maze(GRID_COLS, GRID_ROWS, seed)
        self.grid     = SquareGrid(passages)

        self.player   = Player(_START)
        self.monster  = Monster(_START, Dijkstra(), (220, 70, 50), "Dijkstra")
        self.revealed = set()

        # ── Place reward off the optimal path, ~4 BFS steps from start ──────
        self.reward_pos = _find_reward(self.grid, _START, _END)
        self.reward_collected = False

        # ── Player guide: path via reward if net cost is better ──────────────
        self.guide_path = _compute_reward_path(self.grid, _START, _END, self.reward_pos)

        # ── Monster: straight Dijkstra to end ────────────────────────────────
        self.monster.compute_path(self.grid, _END)

        self._player_turn  = True
        self._finished     = False
        self._show_result  = False
        self._result_btn   = None

        update_fog(self.revealed, _START, self.grid.passages, FOG_REVEAL_RADIUS)

    def handle_event(self, event: pygame.event.Event) -> None:
        # Mouse – result overlay button
        if event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
            if self._show_result and hasattr(self, "_result_btn"):
                if self._result_btn and self._result_btn.collidepoint(event.pos):
                    if "euclidean" not in self.game_data["tomes"]:
                        self.game_data["tomes"].append("euclidean")
                    self._go("MAZE_L3"); return

        if event.type == pygame.KEYDOWN:
            if event.key == pygame.K_ESCAPE:
                self._go("MENU"); return

            if self._show_result:
                if event.key in (pygame.K_RETURN, pygame.K_SPACE):
                    if "euclidean" not in self.game_data["tomes"]:
                        self.game_data["tomes"].append("euclidean")
                    self._go("MAZE_L3")
                return

            if self._finished or not self._player_turn:
                return

            delta = _arrow_to_delta(event.key)
            if delta:
                self._player_move(delta)

    def _player_move(self, delta: tuple) -> None:
        c, r    = self.player.pos
        new_pos = (c + delta[0], r + delta[1])
        if not self.grid.is_passage(self.player.pos, new_pos):
            return

        self.player.move(new_pos)
        update_fog(self.revealed, new_pos, self.grid.passages, FOG_REVEAL_RADIUS)

        # Check reward collection
        if new_pos == self.reward_pos and not self.reward_collected:
            self.reward_collected = True
            self.player.add_bonus(_REWARD_BONUS)
            # Update guide: now route directly to end
            astar = AStarManhattan()
            path, _, _ = astar.find_path(self.grid, new_pos, _END)
            self.guide_path = path[1:3]

        # Update guide next 2 steps
        if not self.reward_collected and self.guide_path:
            # Advance guide to reflect current position
            astar = AStarManhattan()
            path, _, _ = astar.find_path(self.grid, new_pos, _END)
            # If reward still ahead, route via reward
            if self.reward_pos not in [new_pos]:
                self.guide_path = _compute_reward_path(
                    self.grid, new_pos, _END, self.reward_pos
                )
            else:
                self.guide_path = path[1:3]

        if new_pos == _END:
            self._finish_race("Player"); return

        self._player_turn = False
        self.monster.advance()
        if self.monster.pos == _END:
            self._finish_race("Dijkstra"); return
        self._player_turn = True

    def _finish_race(self, winner: str) -> None:
        self._finished   = True
        self._winner     = winner
        self._show_result = True
        if "MAZE_L2" not in self.game_data["completed_levels"]:
            self.game_data["completed_levels"].append("MAZE_L2")

    def update(self, dt: float) -> None:
        pass

    def draw(self) -> None:
        self.screen.fill(UI_BG)

        rp = None if self.reward_collected else self.reward_pos
        draw_square_maze(
            self.screen,
            self.grid.passages,
            self.revealed,
            self.player.pos,
            _START, _END,
            monsters   = [self.monster],
            guide_path = self.guide_path,
            guide_color= REWARD_COLOR,
            reward_pos = rp,
        )

        bonus_txt = "Reward collected! -10 steps" if self.reward_collected else "Gold circle = -10 steps bonus"
        mon_stats = dict(self.monster.stats()); mon_stats["_label"] = "Dijkstra"

        draw_sidebar(
            self.screen,
            level_name    = "Layer 2 – Weighted Reward",
            player_stats  = self.player.stats(),
            monster_stats = mon_stats,
            algo_info     = [
                {"name":"A* + Reward","color":REWARD_COLOR,"desc":"Routes via bonus","formula":"modified g(n)"},
                {"name":"Dijkstra",   "color":DIJKSTRA_COL,"desc":"Ignores reward", "formula":"h=0"},
            ],
            instructions  = _INSTRUCT,
            tomes         = self.game_data["tomes"],
            extra_lines   = [bonus_txt],
        )

        if self._show_result:
            self._draw_result()

    def _draw_result(self) -> None:
        overlay = pygame.Surface((MAZE_AREA_WIDTH, SCREEN_HEIGHT), pygame.SRCALPHA)
        overlay.fill((0, 0, 0, 170))
        self.screen.blit(overlay, (0, 0))
        cx, cy = MAZE_AREA_WIDTH // 2, SCREEN_HEIGHT // 2
        lines = [
            (f"{'You' if self._winner=='Player' else 'Dijkstra'} won!", UI_SUCCESS),
            ("", UI_TEXT),
            (f"Your steps : {self.player.steps}  (reward saved 10 steps!)" if self.reward_collected
             else f"Your steps : {self.player.steps}", ASTAR_MAN_COL),
            (f"Monster    : {self.monster.steps}", DIJKSTRA_COL),
            ("", UI_TEXT),
            ("Detour + Reward = smart weighted A*", UI_SUBTEXT),
            ("Dijkstra can't reason about future bonuses.", UI_SUBTEXT),
            ("", UI_TEXT),
            ("Unlocked: A* Euclidean Technique!", (255,100,100)),
            ("", UI_TEXT),
            ("[Enter] → Layer 3", UI_TEXT),
        ]
        total_h = len(lines) * 26
        y = cy - total_h // 2
        for text, color in lines:
            if text:
                s = font("body").render(text, True, color)
                self.screen.blit(s, (cx - s.get_width() // 2, y))
            y += 26

        # Clickable button
        bw, bh = 200, 40
        btn = pygame.Rect(cx - bw // 2, y + 4, bw, bh)
        self._result_btn = btn
        mx, my = pygame.mouse.get_pos()
        hov = btn.collidepoint(mx, my)
        pygame.draw.rect(self.screen, (50,50,70) if hov else (30,30,50), btn, border_radius=8)
        pygame.draw.rect(self.screen, UI_SUCCESS, btn, width=2, border_radius=8)
        ls2 = font("body").render("Next Level →", True, UI_SUCCESS)
        self.screen.blit(ls2, (btn.x+(bw-ls2.get_width())//2, btn.y+(bh-ls2.get_height())//2))


# ── Helpers ───────────────────────────────────────────────────────────────────

def _find_reward(grid: SquareGrid, start: tuple, end: tuple) -> tuple:
    """Find a cell ~4 BFS steps from start that isn't on the direct optimal path."""
    from src.algorithms import AStarManhattan
    astar = AStarManhattan()
    opt_path, _, _ = astar.find_path(grid, start, end)
    opt_set  = set(opt_path)
    bfs_dist = grid.bfs_distance(start)

    candidates = [
        pos for pos, d in bfs_dist.items()
        if d == 4 and pos not in opt_set and pos != end
    ]
    if not candidates:
        # Fallback: any cell 3-5 steps away, off optimal path
        candidates = [
            pos for pos, d in bfs_dist.items()
            if 2 <= d <= 6 and pos not in opt_set and pos != end
        ]
    if not candidates:
        candidates = [pos for pos in grid.cells() if pos != start and pos != end]

    return random.choice(candidates)


def _compute_reward_path(grid: SquareGrid, start, end, reward) -> list:
    """Return guide highlight: next 2 steps on path start→reward→end."""
    astar = AStarManhattan()
    p1, _, _ = astar.find_path(grid, start, reward)
    p2, _, _ = astar.find_path(grid, reward, end)
    full = p1 + p2[1:]    # merge (remove duplicate reward node)
    return full[1:3]


def _arrow_to_delta(key):
    return {
        pygame.K_UP:(0,-1), pygame.K_DOWN:(0,1),
        pygame.K_LEFT:(-1,0), pygame.K_RIGHT:(1,0),
    }.get(key)

File: web/src/states/layer2State.js
/**
 * layer2State.js — Layer 2: Weighted Reward
 * Turn-based race with a gold gem granting -10 step bonus.
 * Player's guide path routes via the reward; Dijkstra monster ignores it.
 * Mirrors src/states/layer2_state.py
 */

import { posKey, keyToPos }                       from '../utils.js';
import { GRID_COLS, GRID_ROWS, FOG_REVEAL_RADIUS } from '../config.js';
import { generateSquareMaze }                      from '../maze/generator.js';
import { SquareGrid }                              from '../maze/squareGrid.js';
import { AStarManhattan }                          from '../algorithms/astar.js';
import { Dijkstra }                                from '../algorithms/dijkstra.js';
import {
    SCREEN_HEIGHT, MAZE_AREA_WIDTH, SIDEBAR_X, SIDEBAR_WIDTH, CELL_SIZE, C,
    sqCellRect, updateFog,
    drawSquareMaze, drawSidebarBase,
    drawText, drawRoundRect, drawStatBlock,
} from '../renderer.js';

// ── Fixed positions ───────────────────────────────────────────────────────────
const START_K     = posKey(0, GRID_ROWS - 1);
const END_K       = posKey(GRID_COLS - 1, 0);
const REWARD_BONUS = -10;

const MOVE_KEYS = {
    ArrowUp:[0,-1], ArrowDown:[0,1], ArrowLeft:[-1,0], ArrowRight:[1,0],
    w:[0,-1], s:[0,1], a:[-1,0], d:[1,0],
};

// ── Entity factories ──────────────────────────────────────────────────────────

function makePlayer(startK) {
    return {
        posK:      startK,
        steps:     0,
        startTime: performance.now(),
        move(k)       { this.posK = k; this.steps++; },
        addBonus(n)   { this.steps = Math.max(0, this.steps + n); },
        get elapsed() { return (performance.now() - this.startTime) / 1000; },
    };
}

function makeMonster(startK) {
    return {
        posK:          startK,
        color:         C.dijMon,
        name:          'Dijkstra',
        path:          [],
        pathIdx:       0,
        steps:         0,
        nodesExpanded: 0,
        startTime:     performance.now(),

        computePath(grid, endK) {
            const { path, visitedOrder } = new Dijkstra().findPath(grid, this.posK, endK);
            this.path          = path;
            this.pathIdx       = 0;
            this.nodesExpanded = visitedOrder.length;
        },
        advance() {
            if (this.pathIdx + 1 >= this.path.length) return;
            this.pathIdx++;
            this.posK = this.path[this.pathIdx];
            this.steps++;
        },
        get elapsed() { return (performance.now() - this.startTime) / 1000; },
    };
}

// ── Factory ───────────────────────────────────────────────────────────────────

export function createLayer2State(ctx, goTo, gameData) {
    let grid, player, monster, revealed;
    let rewardPosK, rewardCollected;
    let guidePath = [];
    let playerTurn, finished, winner, showResult;
    let anim = 0, mouseX = 0, mouseY = 0;
    let resultBtnRect = null;

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    function enter() {
        const passages = generateSquareMaze(GRID_COLS, GRID_ROWS);
        grid     = new SquareGrid(passages);
        player   = makePlayer(START_K);
        monster  = makeMonster(START_K);
        revealed = new Set();

        monster.computePath(grid, END_K);

        rewardPosK      = _placeReward();
        rewardCollected = false;
        guidePath       = [];

        playerTurn    = true;
        finished      = false;
        winner        = null;
        showResult    = false;
        anim          = 0;
        resultBtnRect = null;

        updateFog(revealed, START_K, grid.passages, FOG_REVEAL_RADIUS);
        _computeGuidePath();
    }

    // ── Reward placement ──────────────────────────────────────────────────────

    function _placeReward() {
        // Optimal path set (A* start → end)
        const { path: optPath } = new AStarManhattan().findPath(grid, START_K, END_K);
        const optSet = new Set(optPath);

        // BFS distances from start
        const bfsDist = grid.bfsDistance(START_K);

        // Prefer cells exactly 4 steps away that are off the optimal path
        let candidates = [];
        for (const [k, d] of bfsDist) {
            if (d === 4 && !optSet.has(k) && k !== START_K && k !== END_K)
                candidates.push(k);
        }
        if (candidates.length) return candidates[Math.floor(Math.random() * candidates.length)];

        // Fallback: 2-6 steps off-path
        for (const [k, d] of bfsDist) {
            if (d >= 2 && d <= 6 && !optSet.has(k) && k !== START_K && k !== END_K)
                candidates.push(k);
        }
        if (candidates.length) return candidates[Math.floor(Math.random() * candidates.length)];

        // Last resort: any cell except start/end
        for (const k of grid.passages.keys()) {
            if (k !== START_K && k !== END_K) return k;
        }
        return START_K; // should never reach here
    }

    // ── Guide path computation ─────────────────────────────────────────────────

    function _computeGuidePath() {
        const astar = new AStarManhattan();
        let full = [];

        if (!rewardCollected) {
            // Route: player → reward → end
            const { path: p1 } = astar.findPath(grid, player.posK, rewardPosK);
            const { path: p2 } = astar.findPath(grid, rewardPosK, END_K);
            if (p1.length > 0 && p2.length > 0) {
                full = [...p1, ...p2.slice(1)];
            }
        } else {
            // Direct: player → end
            const { path } = astar.findPath(grid, player.posK, END_K);
            full = path;
        }

        // Store next 3 steps (skip player's current pos)
        guidePath = full.slice(1, 4);
    }

    // ── Input ─────────────────────────────────────────────────────────────────

    function handleKey(key) {
        if (showResult) {
            if (key === 'Enter' || key === ' ') _advance(); return;
        }
        if (finished || !playerTurn) return;

        const delta = MOVE_KEYS[key];
        if (!delta) return;
        const [c, r] = keyToPos(player.posK);
        const newK   = posKey(c + delta[0], r + delta[1]);
        if (!grid.isPassage(player.posK, newK)) return;

        // Player moves
        player.move(newK);
        updateFog(revealed, newK, grid.passages, FOG_REVEAL_RADIUS);

        // Reward collection
        if (newK === rewardPosK && !rewardCollected) {
            rewardCollected = true;
            player.addBonus(REWARD_BONUS);   // -10 steps
            _computeGuidePath();
        }

        if (newK === END_K) { _endRace('Player'); return; }

        // Monster step
        playerTurn = false;
        monster.advance();
        if (monster.posK === END_K) { _endRace('Dijkstra'); return; }

        _computeGuidePath();
        playerTurn = true;
    }

    function handleMouseMove(x, y) { mouseX = x; mouseY = y; }

    function handleClick(x, y) {
        if (showResult && resultBtnRect && _hit(resultBtnRect, x, y)) _advance();
    }

    function _advance() {
        if (!gameData.tomes.includes('euclidean'))
            gameData.tomes.push('euclidean');
        goTo('MAZE_L3');
    }

    function _endRace(w) {
        finished   = true;
        winner     = w;
        showResult = true;
        if (!gameData.completedLevels.includes('MAZE_L2'))
            gameData.completedLevels.push('MAZE_L2');
    }

    // ── Update ────────────────────────────────────────────────────────────────

    function update(dt) { anim += dt; }

    // ── Draw ──────────────────────────────────────────────────────────────────

    function draw() {
        drawSquareMaze(ctx, grid.passages, revealed, player.posK, START_K, END_K, {
            monsters:    [{ posK: monster.posK, color: monster.color }],
            rewardPosK:  rewardCollected ? null : rewardPosK,
        });
        _drawGuidePath();
        _drawSidebar();
        if (showResult) _drawResultOverlay();
    }

    // ── Guide path highlight ──────────────────────────────────────────────────

    function _drawGuidePath() {
        const pulse = 0.5 + 0.5 * Math.sin(anim * 2.8);
        for (const k of guidePath) {
            if (k === END_K || k === START_K || k === rewardPosK) continue;
            if (!revealed.has(k)) continue;
            const [c, r]      = keyToPos(k);
            const { x, y, w } = sqCellRect(c, r);
            const cx = x + w/2, cy = y + w/2;

            // Soft gold fill
            ctx.fillStyle = `rgba(255,220,30,${(0.22 + 0.10 * pulse).toFixed(2)})`;
            ctx.fillRect(x, y, w, w);

            // Pulsing ring
            ctx.strokeStyle = `rgba(255,240,80,${(0.50 + 0.30 * pulse).toFixed(2)})`;
            ctx.lineWidth   = 2;
            ctx.beginPath();
            ctx.arc(cx, cy, w * 0.30 + 3 * pulse, 0, Math.PI * 2);
            ctx.stroke();
        }
    }

    // ── Sidebar ───────────────────────────────────────────────────────────────

    function _drawSidebar() {
        drawSidebarBase(ctx);
        const sx = SIDEBAR_X + 10, sw = SIDEBAR_WIDTH - 20;
        let y = 14;

        // Title
        ctx.font = 'bold 19px Arial'; ctx.fillStyle = C.uiHeading;
        ctx.textAlign = 'center'; ctx.textBaseline = 'top';
        ctx.fillText('Layer 2 – Weighted Reward', SIDEBAR_X + SIDEBAR_WIDTH / 2, y); y += 28;
        _ln(sx, y, sx + sw, y); y += 10;

        // Turn indicator
        const turnLabel = finished
            ? (winner === 'Player' ? '🏆  You Win!' : '💀  Dijkstra Wins!')
            : playerTurn ? '▶  YOUR TURN' : '⏳  Monster moving…';
        const turnColor = finished
            ? (winner === 'Player' ? C.uiSuccess : C.dijMon)
            : playerTurn ? C.uiWarn : C.uiSubtext;
        drawText(ctx, turnLabel, sx + sw/2, y,
                 { size:14, bold:true, color:turnColor, align:'center' }); y += 26;
        _ln(sx, y, sx + sw, y); y += 10;

        // Reward status
        const rewardLine = rewardCollected
            ? '✓  Reward collected!  −10 steps'
            : '◎  Gold gem = −10 steps bonus';
        drawText(ctx, rewardLine, sx, y,
                 { size:13, color: rewardCollected ? C.uiSuccess : C.reward }); y += 22;

        // Stat blocks
        y = drawStatBlock(ctx, sx, y, sw, 'You  (A* guided)', player.steps, player.elapsed, C.player) + 8;
        y = drawStatBlock(ctx, sx, y, sw, 'Dijkstra Monster', monster.steps, monster.elapsed, C.dijMon) + 12;

        // Algorithm cards
        for (const [name, formula, desc, col] of [
            ['A* + Reward', 'g(n) includes −10 bonus', 'Your guide routes via gem', C.reward],
            ['Dijkstra',    'h = 0, ignores weights',  'Monster: direct to exit',  C.dijkstra],
        ]) {
            drawRoundRect(ctx, sx, y, sw, 58, 5, { fill: C.uiPanel });
            ctx.fillStyle = col;
            ctx.beginPath(); ctx.roundRect(sx, y, 4, 58, 2); ctx.fill();
            drawText(ctx, name,    sx+10, y+ 4, { size:14, bold:true, color:col });
            drawText(ctx, formula, sx+10, y+23, { size:12, color:C.uiHilight });
            drawText(ctx, desc,    sx+10, y+41, { size:11, color:C.uiSubtext });
            y += 62;
        }
        y += 4;

        // Instructions
        _ln(sx, y, sx + sw, y); y += 8;
        for (const line of [
            'Arrow keys / WASD to move',
            'Follow the gold highlights',
            'Collect gem before exit!',
            'ESC → Main Menu',
        ]) {
            drawText(ctx, line, sx, y, { size:12, color:C.uiSubtext }); y += 18;
        }
    }

    // ── Result overlay ────────────────────────────────────────────────────────

    function _drawResultOverlay() {
        ctx.fillStyle = 'rgba(0,0,0,0.72)';
        ctx.fillRect(0, 0, MAZE_AREA_WIDTH, SCREEN_HEIGHT);

        const pw = 580, ph = 430;
        const px = (MAZE_AREA_WIDTH - pw) / 2;
        const py = (SCREEN_HEIGHT   - ph) / 2;
        const winColor = winner === 'Player' ? C.uiSuccess : C.dijMon;

        drawRoundRect(ctx, px, py, pw, ph, 14,
                      { fill: '#14142a', stroke: winColor, lineWidth: 2 });

        // Headline
        const headline = winner === 'Player'
            ? '🏆  You reached the exit first!'
            : '💀  Dijkstra reached the exit first!';
        drawText(ctx, headline, px + pw/2, py + 18,
                 { size:19, bold:true, color:winColor, align:'center' });

        ctx.strokeStyle = C.uiBorder; ctx.lineWidth = 1;
        _ln(px+20, py+52, px+pw-20, py+52);

        // Two-column stats
        const lx = px+30, rx = px+pw/2+10, colW = pw/2-40, statY = py+62;
        drawText(ctx, 'You (guided)',  lx+colW/2, statY, { size:13, bold:true, color:C.player,   align:'center' });
        drawText(ctx, 'Dijkstra AI',  rx+colW/2, statY, { size:13, bold:true, color:C.dijkstra, align:'center' });
        _ln(px+pw/2, statY+18, px+pw/2, statY+110);

        drawText(ctx, 'Steps', px+pw/2, statY+22, { size:12, color:C.uiSubtext, align:'center' });
        drawText(ctx, String(player.steps),  lx+colW/2, statY+38, { size:30, bold:true, color:C.player,   align:'center' });
        drawText(ctx, String(monster.steps), rx+colW/2, statY+38, { size:30, bold:true, color:C.dijkstra, align:'center' });

        drawText(ctx, 'Time', px+pw/2, statY+76, { size:12, color:C.uiSubtext, align:'center' });
        drawText(ctx, `${player.elapsed.toFixed(1)}s`,  lx+colW/2, statY+90, { size:20, bold:true, color:C.player,   align:'center' });
        drawText(ctx, `${monster.elapsed.toFixed(1)}s`, rx+colW/2, statY+90, { size:20, bold:true, color:C.dijkstra, align:'center' });

        _ln(px+20, py+184, px+pw-20, py+184);

        // Reward outcome
        const rewardMsg = rewardCollected
            ? `You collected the −10 bonus!  Net steps: ${player.steps}`
            : 'You missed the reward gem — no bonus applied';
        drawText(ctx, rewardMsg, px+pw/2, py+196,
                 { size:13, bold:true, color: rewardCollected ? C.uiSuccess : C.uiWarn, align:'center' });

        // Pedagogical insight
        let iy = py + 222;
        for (const [text, color] of [
            ['Why does A* route via the gem?',                     C.uiHeading],
            ['Weighted edges change g(n) — A* picks the path',     C.uiSubtext],
            ['with lowest total cost, even if it\'s longer.',       C.uiSubtext],
            ['Dijkstra has no cost preference — it ignores bonus cells.', C.uiSubtext],
        ]) {
            drawText(ctx, text, px+pw/2, iy, { size:12, color, align:'center' }); iy += 20;
        }

        // Unlock badge
        iy += 4;
        drawRoundRect(ctx, px+pw/2-160, iy, 320, 28, 6,
                      { fill: '#2a2800', stroke: C.uiHeading, lineWidth: 1 });
        drawText(ctx, '🔓  Unlocked: A* Euclidean Technique!', px+pw/2, iy+6,
                 { size:13, bold:true, color:C.uiHeading, align:'center' });
        iy += 38;

        // Next Level button
        const btnW = 200, btnH = 42;
        const bx = px + (pw - btnW) / 2, by = iy + 4;
        resultBtnRect = { x:bx, y:by, w:btnW, h:btnH };
        const hov = _hit(resultBtnRect, mouseX, mouseY);
        drawRoundRect(ctx, bx, by, btnW, btnH, 8, {
            fill: hov ? '#1e2e1e' : C.uiPanel,
            stroke: C.uiSuccess, lineWidth: 2,
        });
        drawText(ctx, 'Next Level →', bx+btnW/2, by+btnH/2-9,
                 { size:15, bold:true, color:C.uiSuccess, align:'center' });
        drawText(ctx, 'Enter or click', bx+btnW/2, by+btnH-18,
                 { size:11, color:C.uiSubtext, align:'center' });
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    function _ln(x1, y1, x2, y2) {
        ctx.beginPath(); ctx.moveTo(x1,y1); ctx.lineTo(x2,y2); ctx.stroke();
    }

    return { enter, update, draw, handleKey, handleClick, handleMouseMove };
}

function _hit(r, x, y) {
    return x >= r.x && x <= r.x + r.w && y >= r.y && y <= r.y + r.h;
}
